### ISSUE TITLE
Union of a models cause duplicated validation errors

### ISSUE THREAD

**Issue by LKay** [tiangolo/fastapi#482]
**Describe the bug**
When input data is defined as `Union` type and both models have the same field (or inherit the same base class) response contains duplicated validation errors for the same field.

**To Reproduce**

Let's create models and route:

```python
class Base(BaseModel):
    name: str

class ModelA(Base):
    foo: Optional[str] = None

class ModelB(Base):
    bar: Optional[str] = None

Input = Union[ModelA, ModelB]

@api.post("/foo")
async def foo_post(data: Input = Body(...)):
    pass
```

When request is made with empty json payload `{}`, response contains two errors for field `name` (mesage is `field required`).

**Expected behavior**
In such case only one error of the same type should be returned without duplicates.

**Environment:**
 - OS: macOS
 - FastAPI: 0.35
 - Python: 3.7.4

**Comment by mr-bjerre** [comment]
I have a similar issue when I want to allow for either one file or list of files as model

     file: Union[List[UploadFile], UploadFile] = File(...)

Then when I go to `/docs/` I get this

    Fetch error
    Internal Error /openapi.json

**Comment by dmontagu** [comment]
The original issue is because it is showing you the problems parsing it into each possible choice of the Union. I think there has been some discussion of this over in pydantic; it might be worth discussing there.

Either way, I would recommend upgrading to the latest supported versions of fastapi (0.38.2) and pydantic (0.32.2) in case it has any effect.

**Comment by dmontagu** [comment]
@mr-bjerre Your issue is different; this is not currently supported in FastAPI -- you have to pick one or the other. It has to do with how the openapi schema is generated.

I can't imagine a scenario where you'd want to use `Union[List[UploadFile], UploadFile]` over just `List[UploadFile]` though -- am I missing something?

I think it should be theoretically possible to generate a valid openapi spec representing this (even if FastAPI can't now), but if I understand correctly the actual bytes being sent to the server are the same if you were to send a request for a single `UploadFile` vs. a one-element list. So I'm not sure it would ever make sense to do that.

**Comment by mr-bjerre** [comment]
No I agree the use case is shit.

The reason is that I suspected our data provider to sometimes deliver a single file upload and sometimes multiple files. So I was afraid that we’d have to support that on the same endpoint. 

But it seems that that isn’t the case, luckily. So my part can be closed. Thanks for your answers.

**Comment by tiangolo** [comment]
Yep, all as @dmontagu says. It's validating with the different types you declared, and both have a `name`. You are inheriting models, but the JSON Schemas are independent, and both have a `name`.

Thanks @mr-bjerre for reporting back!
